src/charm.py

```diff
@@ -204,7 +204,7 @@ def _request_certificate(self) -> None:
         signing request (CSR) and inserts it into the `certificates` relation unit relation data.
         """
         if not self._csr_is_stored:
-            raise RuntimeError("CSR is not stored")
+            return
```

Comment:
```suggestion
            return None
```

Reply:
I wouldn't keep the check if it can never be true.
Can this be true?
can `_generate_csr` return before the secret is actually stored in juju?
if so, we should keep the condition, but returning on entering a waiting state might not be recoverable.
I assume that was the idea behind raising an error.

Reply:
I'm not sure why it's beneficial to add the `None` here. Just `return` seems pretty standard and acceptable to me. Is there a reason to add the `None` (it isn't used elsewhere)?

Reply:
I have investigated the flow for the situation that the relation is quickly created by adding extra logging. When the  _request_certificate is called for the first time, csr is not stored. So it raises RuntimeError as follows.

```
unit-tls-certificates-requirer-0: 22:24:40 ERROR juju.worker.uniter.operation hook "config-changed" (via hook dispatching script: dispatch) failed: exit status 1
unit-tls-certificates-requirer-0: 22:24:40 INFO juju.worker.uniter awaiting error resolution for "config-changed" hook
unit-tls-certificates-requirer-0: 22:26:03 INFO juju.worker.uniter awaiting error resolution for "config-changed" hook
unit-tls-certificates-requirer-0: 22:26:05 INFO unit.tls-certificates-requirer/0.juju-log certificate_relation created
unit-tls-certificates-requirer-0: 22:26:05 INFO unit.tls-certificates-requirer/0.juju-log csr is not stored
unit-tls-certificates-requirer-0: 22:26:07 INFO unit.tls-certificates-requirer/0.juju-log private key is stored
unit-tls-certificates-requirer-0: 22:26:09 INFO unit.tls-certificates-requirer/0.juju-log csr generated
unit-tls-certificates-requirer-0: 22:26:09 INFO unit.tls-certificates-requirer/0.juju-log adding secret
unit-tls-certificates-requirer-0: 22:26:09 INFO unit.tls-certificates-requirer/0.juju-log csr generated
unit-tls-certificates-requirer-0: 22:26:09 INFO unit.tls-certificates-requirer/0.juju-log csr is not stored
unit-tls-certificates-requirer-0: 22:26:09 ERROR unit.tls-certificates-requirer/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/./src/charm.py", line 318, in <module>
    main(TLSRequirerOperatorCharm)
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/venv/ops/main.py", line 441, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/venv/ops/main.py", line 149, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/venv/ops/framework.py", line 344, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/venv/ops/framework.py", line 841, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/venv/ops/framework.py", line 930, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/./src/charm.py", line 174, in _on_config_changed
    self._request_certificate()
  File "/var/lib/juju/agents/unit-tls-certificates-requirer-0/charm/./src/charm.py", line 213, in _request_certificate
    raise RuntimeError("CSR is not stored")
RuntimeError: CSR is not stored
```

Within this patch _request_certificate just returns and after start hook when _request_certificate called again, csr is found and it works properly.

```
unit-tls-certificates-requirer-0: 22:57:59 INFO juju.downloader download complete ("local:jammy/tls-certificates-requirer-10")
unit-tls-certificates-requirer-0: 22:57:59 INFO juju.downloader download verified ("local:jammy/tls-certificates-requirer-10")
controller-0: 22:58:02 INFO juju.worker.caasapplicationprovisioner.tls-certificates-requirer scaling application "tls-certificates-requirer" to desired scale 1
unit-tls-certificates-requirer-0: 22:58:13 INFO juju.worker.uniter hooks are retried true
unit-tls-certificates-requirer-0: 22:58:13 INFO juju.worker.uniter found queued "install" hook
unit-tls-certificates-requirer-0: 22:58:14 INFO unit.tls-certificates-requirer/0.juju-log Running legacy hooks/install.
unit-tls-certificates-requirer-0: 22:58:14 INFO unit.tls-certificates-requirer/0.juju-log Private key generated
unit-tls-certificates-requirer-0: 22:58:16 INFO juju.worker.uniter.operation ran "install" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:17 INFO juju.worker.uniter.operation ran "certificates-relation-created" hook (via hook dispatching script: dispatch)
unit-self-signed-certificates-0: 22:58:17 INFO juju.worker.uniter.operation ran "certificates-relation-joined" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:17 INFO juju.worker.uniter found queued "leader-elected" hook
unit-tls-certificates-requirer-0: 22:58:17 INFO juju.worker.uniter.operation ran "leader-elected" hook (via hook dispatching script: dispatch)
unit-self-signed-certificates-0: 22:58:17 INFO juju.worker.uniter.operation ran "certificates-relation-changed" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:19 INFO unit.tls-certificates-requirer/0.juju-log certificate_relation created
unit-tls-certificates-requirer-0: 22:58:19 INFO unit.tls-certificates-requirer/0.juju-log csr is not stored
unit-tls-certificates-requirer-0: 22:58:21 INFO unit.tls-certificates-requirer/0.juju-log private key is stored
unit-tls-certificates-requirer-0: 22:58:24 INFO unit.tls-certificates-requirer/0.juju-log csr generated
unit-tls-certificates-requirer-0: 22:58:24 INFO unit.tls-certificates-requirer/0.juju-log adding secret
unit-tls-certificates-requirer-0: 22:58:24 INFO unit.tls-certificates-requirer/0.juju-log csr generated
unit-tls-certificates-requirer-0: 22:58:24 INFO unit.tls-certificates-requirer/0.juju-log csr is not stored
unit-tls-certificates-requirer-0: 22:58:26 INFO juju.worker.uniter.operation ran "config-changed" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:26 INFO juju.worker.uniter found queued "start" hook
unit-tls-certificates-requirer-0: 22:58:26 INFO unit.tls-certificates-requirer/0.juju-log Running legacy hooks/start.
unit-tls-certificates-requirer-0: 22:58:26 INFO juju.worker.uniter.operation ran "start" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:27 INFO juju.worker.uniter.operation ran "secret-remove" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:33 INFO unit.tls-certificates-requirer/0.juju-log certificates:31: Certificate request sent to provider
unit-tls-certificates-requirer-0: 22:58:33 INFO unit.tls-certificates-requirer/0.juju-log certificates:31: certificate creation requested
unit-tls-certificates-requirer-0: 22:58:33 INFO juju.worker.uniter.operation ran "certificates-relation-joined" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:34 INFO juju.worker.uniter.operation ran "secret-remove" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:34 WARNING unit.tls-certificates-requirer/0.juju-log certificates:31: Provider relation data did not pass JSON Schema validation: {}
unit-tls-certificates-requirer-0: 22:58:34 INFO juju.worker.uniter.operation ran "certificates-relation-changed" hook (via hook dispatching script: dispatch)
unit-self-signed-certificates-0: 22:58:38 INFO unit.self-signed-certificates/0.juju-log certificates:31: Generated certificate for relation 31
unit-self-signed-certificates-0: 22:58:38 INFO juju.worker.uniter.operation ran "certificates-relation-changed" hook (via hook dispatching script: dispatch)
unit-self-signed-certificates-0: 22:58:38 INFO juju.worker.uniter.operation ran "secret-remove" hook (via hook dispatching script: dispatch)
unit-tls-certificates-requirer-0: 22:58:40 INFO unit.tls-certificates-requirer/0.juju-log certificates:31: New certificate is stored: -----BEGIN CERTIFICATE-----
MIIDFDCCAfygAwIBAgIUQpAP+8g9O6i9MxAk1zPuyoecFUQwDQYJKoZIhvcNAQEL
BQAwOTELMAkGA1UEBhMCVVMxKjAoBgNVBAMMIXNlbGYtc2lnbmVkLWNlcnRpZmlj
YXRlcy1vcGVyYXRvcjAeFw0yMzA4MjMxOTU4MzhaFw0yNDA4MjIxOTU4MzhaME8x
HjAcBgNVBAMMFXRscy1yZXF1aXJlci1vcGVyYXRvcjEtMCsGA1UELQwkODM2NTU0
ZjAtNTg4Ni00ZTU4LWIzZTEtODU0NzE3NWYzYzk2MIIBIjANBgkqhkiG9w0BAQEF
AAOCAQ8AMIIBCgKCAQEA9FGznHHC/+XhnUrS8a9W69e6Qf4r/PM02kXDPnjtW8hE
BdaB8t5o4A0b1ysG4VJacElLJrIm0y01dG1lIi90/yY40w8JRLKuaXtC+qsP+SAY
DI+WyHfdpDitaLDNFm+MeZ0LACuUsD7Nngv1fU93hRBp6VAx65VpNIHBtf7Ul50O
VEc0xBTeYwZKRmo78ZS2lNwML0YbWslrbtoBWeuZuCKSfd9elDAlR47kl2CC+ZAB
Km0GfHxtMqNBf8/XlXGr5CfN2kCiNncyUiInOw3M3PV0aV6dAjJJwnyotqZ1+OyR
J90Flb/EtJI1xulZrjTwt0/6cMOZY4ZmvqVgGxvKFwIDAQABMA0GCSqGSIb3DQEB
CwUAA4IBAQAE9Ysw1aM6B9LYWanT4budT++s/psfJYXdwNusnogRZ5NVH3nGUS0x
FHD9iFPTBlA3pIDCWIvVJcgQcslSERc5IaIr+VnEvDREIb6kbZ7kFYal/yICCxqo
EAufPaSOfnS04H+Ws5/nBa/3cG6rYSSNdSeefaXu3EQ/RZwdyYRup7FMX2G1oMJ+
vbc7vZcvFCEi6vpbYxSsQ3HSB33Qj0EOPXf/ZrszZSNpkkfzBlYTl8kQqeWb3PF8
dNt4ftxclLPuZt1VEqYs5d3OZ230VsP1GsM95QoYwu/S9kQUN5CQQ4MPw4JxcWqq
Z46ZV02CPVwE/4O/D65qMrffTvfxf9xp
-----END CERTIFICATE-----
unit-tls-certificates-requirer-0: 22:58:42 INFO juju.worker.uniter.operation ran "certificates-relation-changed" hook (via hook dispatching script: dispatch)
unit-self-signed-certificates-0: 22:58:48 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
```

Reply:
I agree Daniel that using just "return" is fine for this situation.

Reply:
I'm still a bit concerned that some sort of race condition can cause the charm to be stuck here.
But since I couldn't point out that situation this should be good enough for now.

For the "return" it's fine, I see that's how it's being used in the file anyway :)